### ISSUE TITLE
Error in RefreshTokenManager - Support for doctrine/common +3.0

### ISSUE THREAD

**Issue by HeiJon** [markitosgv/JWTRefreshTokenBundle#200]
Since Doctrine Common 3.0, Common was split into smaller packages.

Actually in [RefreshTokenManager](https://github.com/markitosgv/JWTRefreshTokenBundle/blob/master/Doctrine/RefreshTokenManager.php) you inject `Doctrine\Common\Persistence\ObjectManager` in constructor.

Since Doctrine Common 3.0, `Doctrine\Common\Persistence\ObjectManager` became `Doctrine\Persistence\ObjectManager`.

**Comment by byhaskell** [comment]
Hello. Faced the same problem, sharing suddenly slows down the development process.
Need to override manager, my solution is:
```
    gesdinet.jwtrefreshtoken.refresh_token_manager:
        class: App\Doctrine\RefreshTokenManager
        public: true
        arguments: [ '@doctrine.orm.default_entity_manager', 'Gesdinet\JWTRefreshTokenBundle\Entity\RefreshToken']
```

```
<?php
namespace App\Doctrine;

use Doctrine\Persistence\ObjectManager;
use Doctrine\ORM\EntityManagerInterface;
use Gesdinet\JWTRefreshTokenBundle\Entity\RefreshTokenRepository;
use Gesdinet\JWTRefreshTokenBundle\Model\RefreshTokenManager as BaseRefreshTokenManager;
use Gesdinet\JWTRefreshTokenBundle\Model\RefreshTokenInterface;

class RefreshTokenManager extends BaseRefreshTokenManager
{
    /**
     * @var EntityManagerInterface
     */
    protected $objectManager;
    
    /**
     * @var string
     */
    protected $class;
    
    /**
     * @var RefreshTokenRepository
     */
    protected $repository;
    
    /**
     * Constructor.
     *
     * @param ObjectManager $om
     * @param string        $class
     */
    public function __construct(ObjectManager $om, $class)
    {
        $this->objectManager = $om;
        $this->repository = $om->getRepository($class);
        $metadata = $om->getClassMetadata($class);
        $this->class = $metadata->getName();
    }
    
    /**
     * @param string $refreshToken
     *
     * @return RefreshTokenInterface
     */
    public function get($refreshToken)
    {
        return $this->repository->findOneBy(array('refreshToken' => $refreshToken));
    }
    
    /**
     * @param string $username
     *
     * @return RefreshTokenInterface
     */
    public function getLastFromUsername($username)
    {
        return $this->repository->findOneBy(array('username' => $username), array('valid' => 'DESC'));
    }
    
    /**
     * @param RefreshTokenInterface $refreshToken
     * @param bool|true             $andFlush
     */
    public function save(RefreshTokenInterface $refreshToken, $andFlush = true)
    {
        $this->objectManager->persist($refreshToken);
        
        if ($andFlush) {
            $this->objectManager->flush();
        }
    }
    
    /**
     * @param RefreshTokenInterface $refreshToken
     * @param bool                  $andFlush
     */
    public function delete(RefreshTokenInterface $refreshToken, $andFlush = true)
    {
        $this->objectManager->remove($refreshToken);
        
        if ($andFlush) {
            $this->objectManager->flush();
        }
    }
    
    /**
     * @param \DateTime $datetime
     * @param bool      $andFlush
     *
     * @return RefreshTokenInterface[]
     */
    public function revokeAllInvalid($datetime = null, $andFlush = true)
    {
        $invalidTokens = $this->repository->findInvalid($datetime);
        
        foreach ($invalidTokens as $invalidToken) {
            $this->objectManager->remove($invalidToken);
        }
        
        if ($andFlush) {
            $this->objectManager->flush();
        }
        
        return $invalidTokens;
    }
    
    /**
     * Returns the RefreshToken fully qualified class name.
     *
     * @return string
     */
    public function getClass()
    {
        return $this->class;
    }
}
```

**Comment by bcrespy** [comment]
Adding the relevant error for the people coming from google:

`Argument 1 passed to Gesdinet\\JWTRefreshTokenBundle\\Doctrine\\RefreshTokenManager::__construct() must be an instance of Doctrine\\Common\\Persistence\\ObjectManager, instance of Doctrine\\ORM\\EntityManager given, called in \getGesdinet_Jwtrefreshtoken_RefreshTokenManagerService.php on line 24`

**Comment by adityacodes** [comment]
> Hello. Faced the same problem, my solution is:
> 
> ```
>     gesdinet.jwtrefreshtoken.refresh_token_manager:
>         class: App\Doctrine\RefreshTokenManager
>         public: true
>         arguments: [ '@doctrine.orm.default_entity_manager', 'Gesdinet\JWTRefreshTokenBundle\Entity\RefreshToken']
> ```
> 
> ```
> <?php
> namespace App\Doctrine;
> 
> use Doctrine\Persistence\ObjectManager;
> use Doctrine\ORM\EntityManagerInterface;
> use Gesdinet\JWTRefreshTokenBundle\Entity\RefreshTokenRepository;
> use Gesdinet\JWTRefreshTokenBundle\Model\RefreshTokenManager as BaseRefreshTokenManager;
> use Gesdinet\JWTRefreshTokenBundle\Model\RefreshTokenInterface;
> 
> class RefreshTokenManager extends BaseRefreshTokenManager
> {
>     /**
>      * @var EntityManagerInterface
>      */
>     protected $objectManager;
>     
>     /**
>      * @var string
>      */
>     protected $class;
>     
>     /**
>      * @var RefreshTokenRepository
>      */
>     protected $repository;
>     
>     /**
>      * Constructor.
>      *
>      * @param ObjectManager $om
>      * @param string        $class
>      */
>     public function __construct(ObjectManager $om, $class)
>     {
>         $this->objectManager = $om;
>         $this->repository = $om->getRepository($class);
>         $metadata = $om->getClassMetadata($class);
>         $this->class = $metadata->getName();
>     }
>     
>     /**
>      * @param string $refreshToken
>      *
>      * @return RefreshTokenInterface
>      */
>     public function get($refreshToken)
>     {
>         return $this->repository->findOneBy(array('refreshToken' => $refreshToken));
>     }
>     
>     /**
>      * @param string $username
>      *
>      * @return RefreshTokenInterface
>      */
>     public function getLastFromUsername($username)
>     {
>         return $this->repository->findOneBy(array('username' => $username), array('valid' => 'DESC'));
>     }
>     
>     /**
>      * @param RefreshTokenInterface $refreshToken
>      * @param bool|true             $andFlush
>      */
>     public function save(RefreshTokenInterface $refreshToken, $andFlush = true)
>     {
>         $this->objectManager->persist($refreshToken);
>         
>         if ($andFlush) {
>             $this->objectManager->flush();
>         }
>     }
>     
>     /**
>      * @param RefreshTokenInterface $refreshToken
>      * @param bool                  $andFlush
>      */
>     public function delete(RefreshTokenInterface $refreshToken, $andFlush = true)
>     {
>         $this->objectManager->remove($refreshToken);
>         
>         if ($andFlush) {
>             $this->objectManager->flush();
>         }
>     }
>     
>     /**
>      * @param \DateTime $datetime
>      * @param bool      $andFlush
>      *
>      * @return RefreshTokenInterface[]
>      */
>     public function revokeAllInvalid($datetime = null, $andFlush = true)
>     {
>         $invalidTokens = $this->repository->findInvalid($datetime);
>         
>         foreach ($invalidTokens as $invalidToken) {
>             $this->objectManager->remove($invalidToken);
>         }
>         
>         if ($andFlush) {
>             $this->objectManager->flush();
>         }
>         
>         return $invalidTokens;
>     }
>     
>     /**
>      * Returns the RefreshToken fully qualified class name.
>      *
>      * @return string
>      */
>     public function getClass()
>     {
>         return $this->class;
>     }
> }
> ```

Thank You @byhaskell . Added the config to services in config/services.yaml and it worked like a charm.

**Comment by bcrespy** [comment]
I can confirm that the solution provided by @byhaskell works _out of the box_. Thanks !

**Comment by samikroon** [comment]
When will this be merged and released?

**Comment by prathje** [comment]
I hope that this gets merged soon ;)
For the ones using PHPStan (based on @byhaskell):

```
<?php

declare(strict_types=1);

namespace App\Doctrine;

use Doctrine\Persistence\ObjectManager;
use Gesdinet\JWTRefreshTokenBundle\Entity\RefreshTokenRepository;
use Gesdinet\JWTRefreshTokenBundle\Model\RefreshTokenInterface;
use Gesdinet\JWTRefreshTokenBundle\Model\RefreshTokenManager as BaseRefreshTokenManager;

/**
 * Class RefreshTokenManager.
 */
class RefreshTokenManager extends BaseRefreshTokenManager
{
    protected ObjectManager $objectManager;

    /**
     * @var class-string<mixed>
     */
    protected string $class;

    protected RefreshTokenRepository $repository;

    /**
     * Constructor.
     *
     * @param class-string<mixed> $class
     */
    public function __construct(ObjectManager $om, $class)
    {
        $this->objectManager = $om;

        $repo = $om->getRepository($class);
        assert($repo instanceof RefreshTokenRepository);
        $this->repository = $repo;

        $metadata = $om->getClassMetadata($class);

        /** @var class-string<mixed> $classString */
        $classString = $metadata->getName();
        $this->class = $classString;
    }

    /**
     * @param string $refreshToken
     *
     * @return ?RefreshTokenInterface
     */
    public function get($refreshToken): ?RefreshTokenInterface
    {
        $refreshToken = $this->repository->findOneBy(['refreshToken' => $refreshToken]);
        if ($refreshToken instanceof RefreshTokenInterface) {
            return $refreshToken;
        } else {
            return null;
        }
    }

    /**
     * @param string $username
     *
     * @return RefreshTokenInterface
     */
    public function getLastFromUsername($username): ?RefreshTokenInterface
    {
        $refreshToken = $this->repository->findOneBy(['username' => $username], ['valid' => 'DESC']);

        if ($refreshToken instanceof RefreshTokenInterface) {
            return $refreshToken;
        } else {
            return null;
        }
    }

    /**
     * @param bool|true $andFlush
     */
    public function save(RefreshTokenInterface $refreshToken, $andFlush = true): void
    {
        $this->objectManager->persist($refreshToken);

        if ($andFlush) {
            $this->objectManager->flush();
        }
    }

    /**
     * @param bool $andFlush
     */
    public function delete(RefreshTokenInterface $refreshToken, $andFlush = true): void
    {
        $this->objectManager->remove($refreshToken);

        if ($andFlush) {
            $this->objectManager->flush();
        }
    }

    /**
     * @param \DateTime $datetime
     * @param bool      $andFlush
     *
     * @return RefreshTokenInterface[]
     */
    public function revokeAllInvalid(?\DateTimeInterface $datetime = null, $andFlush = true)
    {
        /* @phpstan-ignore-next-line */
        $invalidTokens = $this->repository->findInvalid($datetime);

        foreach ($invalidTokens as $invalidToken) {
            $this->objectManager->remove($invalidToken);
        }

        if ($andFlush) {
            $this->objectManager->flush();
        }

        return $invalidTokens;
    }

    /**
     * Returns the RefreshToken fully qualified class name.
     *
     * @return string
     */
    public function getClass()
    {
        return $this->class;
    }
}
```

**Comment by shehi** [comment]
Yea, stuck with this problem in doctrine/mongodb-odm.

**Comment by kayobruno** [comment]
Same problem :(

**Comment by djpremier** [comment]
Cannot autowire service "App\Doctrine\RefreshTokenManager": argument "$om" of method "__construct()" references interface "Doctrine\Persistence\ObjectManager" but no such service exists. 
You should maybe alias this interface to the existing "doctrine.orm.default_entity_manager" service.

**Comment by Ischci** [comment]
Any idea when this will be fixed?

**Comment by acasademont** [comment]
A simpler approach until this is fixed upstream:

override the service:

```yaml
    gesdinet.jwtrefreshtoken.refresh_token_manager:
      class:  App\Doctrine\RefreshTokenManager
      public: true
      arguments: [ '@gesdinet.jwtrefreshtoken.object_manager', '%gesdinet.jwtrefreshtoken.refresh_token.class%' ]
```

and on the class just override the contructor

```php
<?php

namespace App\Infrastructure\Share\Doctrine;

use Doctrine\Persistence\ObjectManager;

class RefreshTokenManager extends \Gesdinet\JWTRefreshTokenBundle\Doctrine\RefreshTokenManager
{
    public function __construct(ObjectManager $om, $class)
    {
        $this->objectManager = $om;
        $this->repository = $om->getRepository($class);
        $metadata = $om->getClassMetadata($class);
        $this->class = $metadata->getName();
    }
}

**Comment by MrMitch** [comment]
@acasademont if you do that, be careful to use the right namespace in the `class` property of your YAML definition.

In your example, the `class` of you YAML is `App\Doctrine\RefreshTokenManager` and the FQDN of the PHP class you provide is `App\Infrastructure\Share\Doctrine\RefreshTokenManager`.

Those 2 values should match.

**Comment by acasademont** [comment]
Ah yes, sure, I just wanted to change the namespace for the example and I
forgot to change it on the php file, thanks for noticing! The original is
ok!

**Comment by MarlonAEC** [comment]
@byhaskell yeah, excellent that worked for me, thanks!!... I hope the team merges this soon!! Regards!!!

**Comment by Bogro** [comment]
I had the same problem and I confirm that @byhaskell  solution is the right one. Thank you!

**Comment by c7nj7n** [comment]
@acasademont your shorter solution works like charm, thanks and kudos!

**Comment by terox** [comment]
The @acasademont solution works perfectly. Thank you

Waiting the merge :)

**Comment by fd6130** [comment]
Thx for the solution.  I hope the PR will be merge soon.

**Comment by lyrixx** [comment]
Hello @markitosgv 

There are **5** PR to address this issue:

* https://github.com/markitosgv/JWTRefreshTokenBundle/pull/213
* https://github.com/markitosgv/JWTRefreshTokenBundle/pull/209
* https://github.com/markitosgv/JWTRefreshTokenBundle/pull/207
* https://github.com/markitosgv/JWTRefreshTokenBundle/pull/206
* https://github.com/markitosgv/JWTRefreshTokenBundle/pull/204

Could you merge one of theses please ? 

---
And if you don't have time for this, this not a big deal. But Could you give write permissions to someone else? 🙏🏼 

Have a nice day

**Comment by richard-keller** [comment]
Thank you @byhaskell. Worked perfectly.

**Comment by japlavaren** [comment]
hi,

even simpler solution is just configure `RefreshTokenManager` manually without creating custom class

```
gesdinet.jwtrefreshtoken.refresh_token_manager:
        class: Gesdinet\JWTRefreshTokenBundle\Doctrine\RefreshTokenManager
        arguments: [ '@doctrine.orm.default_entity_manager', 'Gesdinet\JWTRefreshTokenBundle\Entity\RefreshToken' ]
```

tested on:
- `"gesdinet/jwt-refresh-token-bundle": "^0.9"`
- `"doctrine/common": "^3.1"`

**Comment by johnnydi** [comment]
Hi, first of all thank you all for your insights. 
I tried the solutions in this thread but got the following ClassNotFound exception in var/cache/dev/ContainerDCJtw8v/getGesdinet_Jwtrefreshtoken_RefreshTokenManagerService.php (line 20):

`     *     * @return \App\Doctrine\RefreshTokenManager     */    
public static function do($container, $lazyLoad = true)    {        
return $container->services['gesdinet.jwtrefreshtoken.refresh_token_manager'] = new \App\Doctrine\RefreshTokenManager(($container->services['doctrine.orm.default_entity_manager'] ?? $container->getDoctrine_Orm_DefaultEntityManagerService()), 'Gesdinet\\JWTRefreshTokenBundle\\Entity\\RefreshToken');   
 }}`
Any ideas? Thank you in Advance

**Comment by samikroon** [comment]
@johnnydi clear the cache should fix that, otherwise remove the contents of the cache folder.

**Comment by markitosgv** [comment]
Solved in v0.10.0, please confirm that!

**Comment by lyrixx** [comment]
Hello @markitosgv
Thanks for taking time to fix this issue.
I confirm the latest release fixed our CI with Symfony 5.2

**Comment by acasademont** [comment]
I can also confirm the issue has been fixed, thanks @markitosgv! I believe this issue could be closed
